pkg/deployer/types.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2022 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deployer
+
+import (
+	"context"
+)
+
+// Deployer is the general interface to handle the deployment of GreptimeDB cluster in different envrionment.
+type Deployer interface {
+	// GetGreptimeDBCluster get the current deployed GreptimeDBCluster by its name.
+	// The name is the namespaced name(<namespace>/<name>) in Kubernetes.
+	GetGreptimeDBCluster(ctx context.Context, name string, options *GetGreptimeDBClusterOptions) (*GreptimeDBCluster, error)
+
+	// ListGreptimeDBClusters list all the deployed GreptimeDBClusters.
+	ListGreptimeDBClusters(ctx context.Context, options *ListGreptimeDBClustersOptions) ([]*GreptimeDBCluster, error)
+
+	// CreateGreptimeDBCluster create a GreptimeDBCluster with the given cluster name.
+	// The name is the namespaced name(<namespace>/<name>) in Kubernetes.
+	CreateGreptimeDBCluster(ctx context.Context, name string, options *CreateGreptimeDBClusterOptions) error
+
+	// UpdateGreptimeDBCluster update the GreptimeDBCluster spec with the given cluster name.
+	// The name is the namespaced name(<namespace>/<name>) in Kubernetes.
```

Comment:
It is better to use namespaceName ？

Reply:
Let's keep API more general. The API of the deployer can run not only on K8s but also on the bare-metal environment. For that reason, `name` would be the better naming.